maven-artifact/src/main/java/org/apache/maven/artifact/versioning/DefaultArtifactVersion.java

```diff
@@ -163,30 +158,42 @@ public final void parseVersion( String version )
             boolean fallback = false;
 
             StringTokenizer tok = new StringTokenizer( part1, "." );
-            try
+            if ( tok.hasMoreTokens() )
             {
                 majorVersion = getNextIntegerToken( tok );
-                if ( tok.hasMoreTokens() )
-                {
-                    minorVersion = getNextIntegerToken( tok );
-                }
-                if ( tok.hasMoreTokens() )
+                if ( majorVersion == null )
                 {
-                    incrementalVersion = getNextIntegerToken( tok );
+                    fallback = true;
                 }
-                if ( tok.hasMoreTokens() )
+            }
+            else
+            {
+                fallback = true;
+            }
+            if ( tok.hasMoreTokens() )
+            {
+                minorVersion = getNextIntegerToken( tok );
+                if ( minorVersion == null )
                 {
-                    qualifier = tok.nextToken();
-                    fallback = Pattern.compile( "\\d+" ).matcher( qualifier ).matches();
+                    fallback = true;
                 }
-
-                // string tokenizer won't detect these and ignores them
-                if ( part1.contains( ".." ) || part1.startsWith( "." ) || part1.endsWith( "." ) )
+            }
+            if ( tok.hasMoreTokens() )
+            {
+                incrementalVersion = getNextIntegerToken( tok );
+                if ( incrementalVersion == null )
                 {
                     fallback = true;
                 }
             }
-            catch ( NumberFormatException e )
+            if ( tok.hasMoreTokens() )
+            {
+                qualifier = tok.nextToken();
+                fallback = Pattern.compile( "\\d+" ).matcher( qualifier ).matches();
```

Comment:
Why not use the char array approach as before?

Reply:
Because it wasn't a hotspot in my tests.

Reply:
Why not, if you know that this works better that with regex.

Reply:
I generally only change things if I can measure they are a problem. Otherwise one can quickly get into a situation where things get worse or the code gets less readable without a benefit.

Reply:
I've adjusted it, as in this case it also improves readability.